crates/engine/src/test_api.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    ext::Engine,
+    types::{
+        AccountId,
+        Balance,
+    },
+    AccountError,
+    Error,
+};
+use std::collections::HashMap;
+
+/// Record for an emitted event.
+#[derive(Clone)]
+pub struct EmittedEvent {
+    /// Recorded topics of the emitted event.
+    pub topics: Vec<Vec<u8>>,
+    /// Recorded encoding of the emitted event.
+    pub data: Vec<u8>,
+}
+
+#[derive(Clone)]
+pub struct RecordedPrintlns {
+    printlns: Vec<String>,
+}
+
+impl RecordedPrintlns {
+    // Creates a new `Engine instance.
+    pub fn new() -> Self {
+        Self {
+            printlns: Vec::new(),
+        }
+    }
+
+    // Records a new println.
+    pub fn record(&mut self, println: String) {
+        self.printlns.push(println);
+    }
+
+    // Clears all recorded printlns.
+    pub fn clear(&mut self) {
+        self.printlns.clear();
+    }
+}
+
+impl Default for RecordedPrintlns {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl IntoIterator for RecordedPrintlns {
+    type Item = String;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.printlns.into_iter()
+    }
+}
+
+/// Recorder for relevant interactions with this crate.
+pub struct DebugInfo {
+    /// Emitted events recorder.
+    emitted_events: Vec<EmittedEvent>,
```

Comment:
Maybe we could do even better by providing an additional emitted event debug sink to the user. (A closure that is called for every emitted event with the event info so the user can do some debug checks themselves in that closure.)

Reply:
As follow-up!